In [ ]:
#hiddencell
from pbl_tools import *

# 스테이지 4

# 1. pandas를 이용해 csv 파일 읽어오기
[문제 1]    
Pandas 라이브러리(library)를 가져와보세요.     
그리고 train.csv, test.csv, sample_submission.csv   파일을 각각 train, test, submission 변수로 읽어오세요.      
아래 빈칸을 채워주세요.

In [ ]:
import ___ as pd

train = pd.___('train.csv')
test = pd.___('test.csv')
submission = pd.___('sample_submission.csv')

In [ ]:
#checkcode
ensure_vals(globals(), 'train', 'test', 'submission')
@check_safety
def check(
          user_answer_train=train,
          user_answer_test=test, 
          user_answer_submission=submission
          ):
    
    c_point0 = hasattr(user_answer_train, 'tail')
    c_point1 = hasattr(user_answer_test, 'tail')
    c_point2 = hasattr(user_answer_submission, 'tail')

    if c_point0 and c_point1 and c_point2 :
        return True
    else:
        return False

check()

True

### Inst.  


### Hint.
`pd.read_csv('파일경로/파일명')` 를 활용해 보세요!

### Solution.
```python
import pandas as pd  

train = pd.read_csv('train.csv')  
test = pd.read_csv('test.csv')  
submission = pd.read_csv('sample_submission.csv')
```

# 2. train 데이터 결측치 확인하기(isna) 

[문제 2]     
`isna()` 메소드를 이용해서 train 데이터셋의 결측치를 확인해 보세요.

In [ ]:
train.___()

title  distributor  genre  release_time   time  screening_rat  director  \
0    False        False  False         False  False          False     False   
1    False        False  False         False  False          False     False   
2    False        False  False         False  False          False     False   
3    False        False  False         False  False          False     False   
4    False        False  False         False  False          False     False   
..     ...          ...    ...           ...    ...            ...       ...   
595  False        False  False         False  False          False     False   
596  False        False  False         False  False          False     False   
597  False        False  False         False  False          False     False   
598  False        False  False         False  False          False     False   
599  False        False  False         False  False          False     False   

     dir_prev_bfnum  dir_prev_num  num_staff  num_actor  box_off_num  
0              True         False      False      False        False  
1             False         False      False      False        False  
2             False         False      False      False        False  
3             False         False      False      False        False  
4             False         False      False      False        False  
..              ...           ...        ...        ...          ...  
595           False         False      False      False        False  
596           False         False      False      False        False  
597            True         False      False      False        False  
598            True         False      False      False        False  
599            True         False      False      False        False  

[600 rows x 12 columns]

In [ ]:
#checkcode
ensure_vals(globals(),'_')
@check_safety
def check(user_answer=_,column='dir_prev_bfnum'):
    c_point0 = user_answer[column][0] == 1
    if c_point0:
        return True
    else:
        return False
check()

### Inst.

`isna()` 함수는 pandas 데이터프레임(DataFrame)에서 **결측값(missing value)을 찾는** 함수입니다.   
이 함수는 데이터프레임의 각 셀에 대하여 결측값 여부를 확인하며, **결측값이 존재하는 경우 True, 그렇지 않은 경우 False** 값을 반환합니다.   
이를 통해 데이터프레임에 포함된 **결측값의 위치**를 파악할 수 있습니다.

**결측값** 은 데이터셋에서 측정되지 않거나, 기록되지 않은 값으로, 이를 그대로 사용하면 모델의 성능을 저하시킬 수 있습니다.   
따라서 결측값이 있는 데이터를 사용하기 전에는 반드시 이를 **처리하는 과정** 이 필요합니다.

해당 코드는 train이라는 pandas 데이터프레임에서 **결측값 여부를 확인** 하는 코드입니다.   
이를 통해 train 데이터프레임에서 **결측값이 있는 열(column)들을 파악** 할 수 있습니다.   
이는 모델의 학습에 결측값이 영향을 미치는 경우 데이터 전처리를 수행할 필요가 있으므로,   
데이터의 결측값 여부를 파악하는 것은 데이터 분석에서 매우 중요한 단계 중 하나입니다. 

---

#### 결과 해석

`dir_prev_bfnum` 피처(feature)에만 `True` 값이 존재하는 것으로 보입니다.   
이는 **`dir_prev_bfnum` 피처에 결측값이 존재한다** 는 것을 의미합니다.   
이러한 결측값이 존재하는 경우, 해당 피처를 모델의 입력으로 사용하기 위해서는 결측값을 대체하는 등의 데이터 전처리 작업이 필요합니다.

<img width= 1100 src="https://drive.google.com/uc?id=19ODu8FH4beJgRzTQMddZYZH5zy1R2fVG">

### Hint.
empty

### Solution.
```python
train.isna()
```

# 3. train 데이터 결측치 확인하기 (isnull)
[문제 3]     
`isnull()` 메소드를 이용해서 train 데이터셋의 결측치를 확인해 보세요.

In [ ]:
train.___()

title  distributor  genre  release_time   time  screening_rat  director  \
0    False        False  False         False  False          False     False   
1    False        False  False         False  False          False     False   
2    False        False  False         False  False          False     False   
3    False        False  False         False  False          False     False   
4    False        False  False         False  False          False     False   
..     ...          ...    ...           ...    ...            ...       ...   
595  False        False  False         False  False          False     False   
596  False        False  False         False  False          False     False   
597  False        False  False         False  False          False     False   
598  False        False  False         False  False          False     False   
599  False        False  False         False  False          False     False   

     dir_prev_bfnum  dir_prev_num  num_staff  num_actor  box_off_num  
0              True         False      False      False        False  
1             False         False      False      False        False  
2             False         False      False      False        False  
3             False         False      False      False        False  
4             False         False      False      False        False  
..              ...           ...        ...        ...          ...  
595           False         False      False      False        False  
596           False         False      False      False        False  
597            True         False      False      False        False  
598            True         False      False      False        False  
599            True         False      False      False        False  

[600 rows x 12 columns]

In [ ]:
#checkcode
ensure_vals(globals(),'_')
@check_safety
def check(user_answer=_,column='dir_prev_bfnum'):
    c_point0 = user_answer[column][0] == 1
    if c_point0:
        return True
    else:
        return False
check()

### Inst.

`isnull()`은 pandas 데이터프레임에서 **결측값(missing value)을 찾는** 함수 중 하나입니다.   
결측값이 존재하는 경우 True, 그렇지 않은 경우 False 값을 반환합니다.

**isnull() 함수는 isna() 함수와 동일한 역할을 수행** 하며, pandas 데이터프레임에서 결측값을 찾는 함수 중 하나입니다.    
결측값을 찾는 것은 모델의 성능에 영향을 미칠 수 있으므로, 결측값을 처리하기 위한 데이터 전처리 작업이 필요합니다.

<img width= 1100 src="https://drive.google.com/uc?id=19ODu8FH4beJgRzTQMddZYZH5zy1R2fVG">

### Hint.
empty

### Solution.
```python
train.isnull()
```

# 4. train 데이터 결측치 확인하기 (notnull)

[문제 4]  
`notnull()` 함수를 활용하여 결측치가 아닌 경우 True의 값을 반환하고,   
결측치인 경우 False의 값을 반환하도록 코드를 작성해 보세요.  
아래 빈칸을 채워주세요.

In [ ]:
train.___()

title  distributor  genre  release_time  time  screening_rat  director  \
0     True         True   True          True  True           True      True   
1     True         True   True          True  True           True      True   
2     True         True   True          True  True           True      True   
3     True         True   True          True  True           True      True   
4     True         True   True          True  True           True      True   
..     ...          ...    ...           ...   ...            ...       ...   
595   True         True   True          True  True           True      True   
596   True         True   True          True  True           True      True   
597   True         True   True          True  True           True      True   
598   True         True   True          True  True           True      True   
599   True         True   True          True  True           True      True   

     dir_prev_bfnum  dir_prev_num  num_staff  num_actor  box_off_num  
0             False          True       True       True         True  
1              True          True       True       True         True  
2              True          True       True       True         True  
3              True          True       True       True         True  
4              True          True       True       True         True  
..              ...           ...        ...        ...          ...  
595            True          True       True       True         True  
596            True          True       True       True         True  
597           False          True       True       True         True  
598           False          True       True       True         True  
599           False          True       True       True         True  

[600 rows x 12 columns]

In [ ]:
#checkcode
def check(user_answer=_):
    train_check = pd.read_csv('train.csv')
    train_check = train.notnull()
    
    if (user_answer == train_check).all().all():
        return True
    else:
        return False
check()

True

### Inst.

`notnull()`은 pandas 데이터프레임(dataframe)에서 **결측값(missing value)이 아닌** 데이터를 찾는 함수입니다.   
`isnull()` 함수와 **반대로 동작**하며, 결측값이 아닌 경우 True, 결측값인 경우 False 값을 반환합니다.

`train.notnull()` 코드는 train 이라는 pandas 데이터프레임에서 결측값이 아닌 데이터를 확인하는 코드입니다.   
이를 통해 train 데이터프레임에서 결측값이 아닌 열(column)들을 파악할 수 있습니다. 

<img width=1200 src="https://drive.google.com/uc?id=1nR1ENaveODIuEdtK1Mp9vfIFE7ksTzk_">

### Hint.
empty

### Solution.
```python
train.notnull() 
```

# 5. 피처(feature)별 결측치 개수 확인

[문제 5]  
`isnull().sum()` 메소드를 사용하여 pandas 데이터프레임에서 **결측값(missing value)의 개수**를 확인해보세요.

In [ ]:
train.___().___()

title               0
distributor         0
genre               0
release_time        0
time                0
screening_rat       0
director            0
dir_prev_bfnum    330
dir_prev_num        0
num_staff           0
num_actor           0
box_off_num         0
dtype: int64

In [ ]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer='_'):
    check_train = pd.read_csv('train.csv')
    check_train = check_train.notnull().sum().sum()
    
    c_point0 = (sum(_) == check_train).all
    if c_point0:
        return True
    else:
        return False
check()

### Inst.

`isnull().sum()` 코드는 pandas 데이터프레임에서 **결측값(missing value)의 개수**를 확인하는 코드입니다. 
 
`isnull()` 함수를 통해 데이터프레임의 각 셀에 대해 **결측값 여부를 확인** 하고,   
이후 `sum()` 함수를 이용하여 각 **열(column)별로 결측값의 총합** 을 계산합니다.

결측값이 많은 열(column)을 찾아 결측값을 처리하거나,   
해당 열(column)을 제거하는 등의 데이터 전처리 작업을 수행할 때 유용합니다.   

---

#### 결과 해석

`dir_prev_bfnum` 열(column)에 330개의 결측치가 있음을 확인할 수 있습니다.   
이러한 결측치는 데이터 분석에서 처리해주어야 하며,   
해당 열(column)을 사용할 경우 결측치를 처리하거나 해당 열(column)을 제거하는 등의 전처리 작업이 필요합니다.






### Hint.
empty

### Solution.
```python
train.isnull().sum()
```

# 6. train 데이터의 결측치가 아닌 값의 개수 확인

[문제 6]  
각 피처(feature)별 결측치가 **아닌** 데이터의 개수를 구해 보세요.  
아래 빈칸을 채워보세요.

In [ ]:
train.___().___()

title             600
distributor       600
genre             600
release_time      600
time              600
screening_rat     600
director          600
dir_prev_bfnum    270
dir_prev_num      600
num_staff         600
num_actor         600
box_off_num       600
dtype: int64

In [ ]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer='_'):
    check_train = pd.read_csv('train.csv')
    check_train = check_train.notnull().sum().sum()
    
    c_point0 = (sum(_) == check_train).all
    if c_point0:
        return True
    else:
        return False
check()

True

### Inst.

`notnull()` 함수는 데이터프레임(DataFrame)의 각 셀에 대해 결측값 여부를 확인하고,   
이후 `sum()` 함수를 이용하여 각 열(column)별로 결측값이 아닌(non-missing) 데이터의 총합을 계산합니다.

`notnull().sum()`을 이용하면 결측치는 False로,   
결측치가 아니면 True로
하여 각 피처(feature)별 결측치가 아닌 데이터의 개수를 구할 수 있습니다.

### Hint.
empty

### Solution.
```python
train.notnull().sum()
```

# 7. 데이터프레임(DataFrame) 정보 출력하기

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           600 non-null    object 
 1   distributor     600 non-null    object 
 2   genre           600 non-null    object 
 3   release_time    600 non-null    object 
 4   time            600 non-null    int64  
 5   screening_rat   600 non-null    object 
 6   director        600 non-null    object 
 7   dir_prev_bfnum  270 non-null    float64
 8   dir_prev_num    600 non-null    int64  
 9   num_staff       600 non-null    int64  
 10  num_actor       600 non-null    int64  
 11  box_off_num     600 non-null    int64  
dtypes: float64(1), int64(5), object(6)
memory usage: 56.4+ KB


In [ ]:
#checkcode
#empty

### Inst.

`info()` 메소드는 pandas 라이브러리(Library)에서 제공하는 함수 중 하나로, **데이터프레임(DataFrame)의 정보를 제공** 합니다.  
데이터프레임의 행(row) 수, 칼럼(column) 수, 각 칼럼의 이름, 데이터 타입, 메모리 사용량 등을 보여주며,  
결측치(null)의 개수도 확인할 수 있습니다.  

이 코드 기법은 데이터 분석 및 전처리 과정에서 매우 유용하게 사용됩니다.  
데이터프레임의 크기와 구성을 파악하여 결측치 처리, 데이터 타입 변경, 이상치 처리 등의 작업을 수행할 수 있습니다.  
또한, 메모리 사용량을 파악하여 데이터프레임의 크기를 최적화하는데도 활용됩니다.  

`info()` 메소드의 출력 결과 중에서, 각 칼럼에 대한 정보는 다음과 같은 의미를 가집니다.  

- **column**: 데이터프레임의 **컬럼(열)** 이름을 보여줍니다.  
- **Non-null count**: 각 컬럼에 있는 **비어있지 않은 값의 개수** 를 보여줍니다.  
- **dtype**: 각 컬럼의 **데이터 타입** 을 보여줍니다. 예를 들어, object는 문자열, int는 정수, float는 소수점이 있는 숫자 등을 의미합니다.  
  
---

#### 결과 해석

train 데이터셋은 총 12개의 열(column)과 600개의 행(row)으로 이루어져 있습니다.    
또한 데이터프레임의 Non-Null Count를 살펴보면 dir_prev_bfnum 변수에 결측치(누락된 데이터)가 330개 존재하는 것을 알 수 있습니다.       
이러한 정보를 바탕으로 데이터프레임의 구성과 각 열의 정보를 파악할 수 있으며, 이는 데이터 전처리나 모델링에 필요한 정보를 제공할 수 있습니다.    

### Hint.
empty

### Solution.
empty